### PR TITLE
Lower case

### PR DESCRIPTION
**Pull request by rgwch**
This makes two changes:
(1) user names are case independent. It is not only unnecessary but also error prone to have "Username" and "username" refer to different users. So, this patch converts all usernames to lowercase in the first place and then also converts usernames to lowercase on create and on login.

(2) Insert a Column "Zeit" into elexis.behandlungen. The new tarmed revision rises the need to create several consultations on the same day and ordering and managing consultations is easier if creation time is visible in the database. (This is not the same as modification time of the consultation entry as given in the 'eintrag' VersionedResource.)

The patch sets the creation time for all newly created consultations to the current time. Existing consultations are not touched. The Comparator will respect date and time, and the "PatHeuteView" orders consultations by date and time now.

### CODE REVIEW

File: ch.elexis.core.data/src/ch/elexis/core/data/util/DBUpdate.java [hunk 1]
```diff
@@ -431,7 +432,11 @@
 			// 3.4.1
 			FILE_LOCATED,
 			// 3.4.2
-			FILE_LOCATED
+			FILE_LOCATED,
+			// 3.4.3
+			"UPDATE user_ SET `id` = LOWER(`id`);"+
+			ALTER_TABLE+Konsultation.TABLENAME+ADD+Konsultation.FLD_TIME+" CHAR(6);"
```
Comment (ngiger) [inline ch.elexis.core.data/src/ch/elexis/core/data/util/DBUpdate.java]: I object about addding a CHAR field with time info.
Instead I suggest adding support for proper handling a SQL datetime column (ISO/IEC 9075-11:2)
See 
* https://dev.mysql.com/doc/refman/5.7/en/date-and-time-literals.html
* https://www.postgresql.org/docs/9.1/static/datatype-datetime.html
* (H2) only has timestamp: http://h2database.com/html/datatypes.html#date_type

The standards are availble (for a too big for me price) under https://www.iso.org/search/x/query/SQL.

**Comment by rgwch** [comment]
The link you provided says:

 MySQL recognizes TIME values in these formats:

    As a string in 'D HH:MM:SS' format. You can also use one of the following “relaxed” syntaxes: 'HH:MM:SS', 'HH:MM', 'D HH:MM', 'D HH', or 'SS'. Here D represents days and can have a value from 0 to 34.

    As a string with no delimiters in 'HHMMSS' format, provided that it makes sense as a time. For example, '101112' is understood as '10:11:12', but '109712' is illegal (it has a nonsensical minute part) and becomes '00:00:00'.

    As a number in HHMMSS format, provided that it makes sense as a time. For example, 101112 is understood as '10:11:12'. The following alternative formats are also understood: SS, MMSS, or HHMMSS. 


So, HHMMSS is perfectly legal for mysql. PostgreSQL prefers obviously HH:MM:SS which is also legal for mysql. So, char(8) would probably be better. To me, the most important thing is, that date and time fields must be readable with all important languages, especially JavaScript, which becomes more and more important for elexis. So, I think the type of the field should always be CHAR or VARCHAR (which is in almost all tables of elexis true for any date or time fields)

**Comment by rgwch** [comment]
I changed the patch to use standard SQL format HH:MM:SS

**Comment by col-panic** [comment]
Both suggested changes are relevant for consideration. Concerning the case-sensitive usernames there is something itching in my brain telling me this is problematic - have to do a redmine search on this. 

Considering the addition of time information to consultation. This is already part of the roadmap for 3.5. If you provide PRs for the project, please do not mix up two separate patches into a single one. This makes matters more difficult to separate - discuss upon - and partially accept!

**Comment by col-panic** [comment]
The itch: https://redmine.medelexis.ch/issues/5506

**Comment by rgwch** [comment]
You can still just convert the username to lowercase at login (as my patch does). So the Username is both, case insensitive and always lowercase. That way I and RH (whatever that might be) should both be happy.

**Comment by rgwch** [comment]
Sorry for mixing two matters in one PR. I first implemented them for myself and then thought later, it might be of general interest. But then it was to difficult to separate them again.

Until 2017, if a patient called 2 times per phone and dropped then in, you could just add time to a single consultation. 2018 this must be separate consultations for limitation reasons.

And: With the change to InnoDB tables, the "PatHeuteView" does not anymore sort consultations automagically by creation time. They come up wildly mixed, thus making it more difficult to control billings. I think adding a Field to consultation does not harm anybody, and CHAR is always more database and language agnostic than some of the many different time/timestamp/timedate fields of databases. CHAR is easy to read and easy to sort with any frontend, and yet compatible to any backend.

If you don't like it, I would suggest at least to skip 3.4.3 on your next database update, so people using ungrad can stay compatible.

**Comment by col-panic** [comment]
As for the case sensitive login - please leave me some time to think about this! I am currently heavy involved in elexis-server authz implementation.

The time patch for consultation is on our agenda (see https://redmine.medelexis.ch/issues/3750) we had a similar case with medication and i would like to compare the solutions here.

A separation of PRs would be very much welcomed here, you could possibly check out the commit with time patch only and then create a PR out of this?!

**Comment by huthomas** [comment]
The case with medication is more a how not to do it. There we extended the time field, and had troubles with the queries because of that. I would prefer the proposed solution using a separate field for the time.

Concerning the format, we should first test if the `:` make a difference for queries involving greater or smaller than on the field.

**Pull request closed without merging**